### ISSUE TITLE
Fonts from FT_Face instead of ttf file

### ISSUE THREAD

**Issue by doctormo** [jpakkane/capypdf#65]
Inkscape has the PangoFont and FT_Face objects for the in-use text objects, but doesn't have filenames. I see from the code that capypdf is creating FT_Font objects. The question is, is a FT_Font object enough information for capypdf to build text or is more information from the ttf file needed?

**Comment by jpakkane** [comment]
Sadly it does not. To do subsetting you have to parse the original font data files and reassemble them. IIRC this is even mentioned in Freetype's own docs.

Even if it were possible, the public API can't really have FT_Face objects in them as they are an implementation detail.

**Comment by doctormo** [comment]
This is a hard red line, there's nothing in FreeType, Pango, Harfbuz or anything else that can route from a font family to a ttf filename. It's not expected usage to have to feed a true type font file into an upstream API in any of the example code on github or gnome.

Basically cairo takes a Pango Font Description and is able to output PDF files from that. With more research I've found that because these font libraries are extremely resistant to giving over information about their font files, I may have no choice but to make capypdf take *something* or give up on capypdf. Which I'm not willing to contemplate yet. Everything else we've managed to solve for, so this must be solvable too.

**Comment by jpakkane** [comment]
I guess they use Fontconfig internally? At least `fc-list -v` shows output like this:

```
Pattern has 28 elts (size 32)
        family: "Noto Serif"(s)
        familylang: "en"(s)
        style: "Regular"(s)
        stylelang: "en"(s)
        fullname: "Noto Serif Regular"(s)
        fullnamelang: "en"(s)
        slant: 0(i)(s)
        weight: 80(f)(s)
        width: 100(f)(s)
        foundry: "GOOG"(s)
        file: "/usr/share/fonts/truetype/noto/NotoSerif-Regular.ttf"(s)
        index: 0(i)(s)
        outline: True(s)
        scalable: True(s)
...
```

**Comment by doctormo** [comment]
I'm scouring the fc codebase, because the property must be there somewhere.

**Comment by jpakkane** [comment]
Maybe grep fontconfig symbol names in Pango or Cairo? Or maybe Inkscape already uses fc to get the font list it displays and it could be made to get the file name at the same time (just speculating)?

In any case you can verify that it does the lookup via the command line tool:

```
$ fc-match courier:italic
NimbusMonoPS-Italic.otf: "Nimbus Mono PS" "Italic"
```

**Comment by doctormo** [comment]
It's because you have to get around pango/hb but not for the FCFont, which seems useless, but for the FcPattern which has an FC_FILE property. At least that's my working theory.

**Comment by jpakkane** [comment]
I experimented with this a bit and the following worked for me:

```c++
#include<fontconfig/fontconfig.h>
#include<cassert>
#include <cstdio>

int main(int argc, char **argv) {
    auto *config = FcInitLoadConfigAndFonts();
    assert(config);
    const char *query = "Palatino";
    if(argc > 1) {
        query = argv[1];
    }
    FcPattern *pat = FcNameParse((const unsigned char*)query);
    FcConfigSubstitute(config, pat, FcMatchPattern);
    FcResult result;
    FcDefaultSubstitute(pat);
    FcPattern *font = FcFontMatch(config, pat, &result);
    assert(font);
    char *path = nullptr;
    auto matchres = FcPatternGetString(font, FC_FILE, 0, (unsigned char**)(&path));
    assert(matchres == FcResultMatch);
    printf("Query: %s\nMatched: %s\n", query, path);
    FcPatternDestroy(font);
    FcPatternDestroy(pat);
    FcConfigDestroy(config);
    FcFini();
    return 0;
}
```

Which results in:

```
Query: Palatino
Matched: /usr/share/fonts/opentype/urw-base35/P052-Roman.otf
```

But I feel that this functionality should be in the calling application, because Fontconfig does font substitution. A PDF generation library should not have heuristics like these. Silently changing the used font from an "expected" one to something else is a _bad_ thing to do.

**Comment by doctormo** [comment]
Agreed, this is already in Inkscape. I have the font file in pretty much the same kind of way. Working through the libnrtype nonsense right now that's entirely built around cairo.
